js/boaexchange.js

```diff
@@ -0,0 +1,1021 @@
+                'Address is Pending': AddressPending,
+                'Insufficient Funds': InsufficientFunds,
+                'Order Not Found': OrderNotFound,
+                'Too Many Requests': DDoSProtection,
+            },
+            'options': {
+                // price precision by quote currency code
+                'pricePrecisionByCode': {
+                    'USD': 3,
+                },
+                'symbolSeparator': '_',
+            },
+        });
+    }
+
+    nonce () {
+        return this.seconds ();
+    }
+
+    async createDepositAddress (code, params = {}) {
+        await this.loadMarkets ();
+        const currency = this.currency (code);
+        const request = {
+            'label': currency['id'],
+        };
+        const response = await this.privateGetAddressesLabel (this.extend (request, params));
+        if (response['data'].length > 0) {
+            return this.parseDepositAddress (response['data'][0]);
+        }
+        return [];
+    }
+
+    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
+        if (type !== 'limit')
+            throw new ExchangeError (this.id + ' allows limit orders only');
+        await this.loadMarkets ();
+        let market = this.market (symbol);
+        let order = {
+            'market': market['id'],
+            'amount': this.amountToPrecision (symbol, amount),
+            'price': this.priceToPrecision (symbol, price),
+            'side': side,
+        };
+        let response = await this.privateGetOrder (this.extend (order, params));
+        return this.extend (this.parseOrder (response['data'], market), {
+            'status': 'open',
+            'price': order['price'],
+            'symbol': symbol,
+            'amount': order['amount'],
+            'side': side,
+            'type': type,
+            'id': response['data']['id'],
+        });
+    }
+
+    async cancelOrder (id, symbol = undefined, params = {}) {
+        await this.loadMarkets ();
+        let request = { 'orderId': id };
+        let response = await this.privateDeleteOrdersOrderId (this.extend (request, params));
+        return this.extend (this.parseOrder (response), {
+            'status': 'canceled',
+        });
+    }
+
+    async cancelOrders (ids, symbol = undefined, params = {}) {
+        await this.loadMarkets ();
+        let request = { 'orderId': ids.join (',') };
+        let response = await this.privateDeleteOrdersOrderId (this.extend (request, params));
+        return this.extend (this.parseOrder (response), {
+            'status': 'canceled',
+        });
+    }
+
+    async fetchBalance (params = {}) {
+        await this.loadMarkets ();
+        let response = await this.privateGetBalances (params);
+        let balances = response['data'];
+        return this.parseBalances (balances);
+    }
+
+    async fetchClosedOrders (symbol = undefined, since = 0, limit = 0, params = {}) {
```

Comment:
Can you do the method signatures exactly how they are defined in the Manual and in the base class? Because otherwise we're introducing a point for a break if we add the next language and so on. See how these checks are done in `hitbtc2` or `binance`, here's a list of rules, basically:

- the arguments are always undefined or required (don't have a default value)
- and if the exchange doesn't require them – we don't send them at all
- If they are required, and the exchange doesn't specify the default value (or it is personalized) – then we throw an `ArgumentsRequired` exception
- if the exchange requires them and specifies a default value (non-personalized parameter) – we can set the exchange default and send the request

This is the behaviour of most other exchanges. There are very persistent users who will send `null` and `undefined` there even if the default is `0`, therefore our default is always `undefined` and the checks are all done inside the body.

Reply:
done. undefined is accepted by the api so that's fine as a default